design/design-doc-agent.md

```diff
@@ -0,0 +1,665 @@
+   --set apiGroups=database.acmeco.org,network.acmeco.org \
+   --set default-credentials-secret-name=agent1-creds \
+   --set default-target-namespace=bar
+   ```
+
+At this point, the setup has been completed. Now, users can use the APIs in the
+remote cluster just as if they are in the local mode. An example application to
+deploy:
+
+```yaml
+apiVersion: common.crossplane.io/v1alpha1
+kind: MySQLInstanceRequirement
+metadata:
+  name: sqldb
+  namespace: default
+spec:
+  version: "5.7"
+  storageGB: 20
+  writeConnectionSecretToRef:
+    name: sql-creds
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: wp
+  namespace: default
+spec:
+containers:
+- name: wordpress
+    image: "wordpress:4.6.1-apache"
+    env:
+    - name: WORDPRESS_DB_PASSWORD
+    valueFrom:
+        secretKeyRef:
+          name: sql-creds
+          key: password
+```
+
+The `MySQLInstanceRequirement` will be synced to `bar` namespace in the central
+cluster. In case there are other `MySQLInstanceRequirement` custom resources in
+the remote cluster with same name but in different namespaces, then the agent
+will reject syncing that in order to prevent the conflict.
+
+
+Note that these steps show the bare-bones setup. Most of the steps can be made
+easier with simple commands in Crossplane CLI and YAML templates you can edit &
+use.
+
+Users can discover what resources available to them and how they
+can consume them from their remote cluster.
+
+List the published APIs:
+```bash
+kubectl get infrastructurepublications
+```
+
+See what keys are included in the connection `Secret` of a specific API so that
+they know what keys to use in mounting process:
+```
+kubectl get infrastructuredefinition mysqlinstance.database.acmeco.org -o yaml
+```
+
+See what `Composition`s are available to select from:
+```
+kubectl get compositions
+```
+
+## Future Considerations
+
+### Additional Crossplane CLI Commands
+
+Crossplane CLI can have simple commands to do most of the setup. For example,
+with one command it should be possible to create the `ServiceAccount` in the
+central cluster together with all of its RBAC resources. Also, agent
+installation could be a very smooth process if we use Crossplane CLI instead of
+Helm.
+
+### RBAC Controller
+
+A controller with its own types to manage the `ServiceAccount`s in the central
+cluster could be a boost to UX. You'd create a custom resource that will result
+in all RBAC resources that are needed for the agent to work with all the APIs in
+the central cluster and write the credentials to a secret. Then the user can
+copy this secret into their remote cluster and refer to it.
+
+### Dynamic Updates to Agent's Role
+
+In the remote cluster, we provide the `Role` that has the static set of
+`apiGroup`s we'd like agent to be able to manage in the remote cluster. There
+could be a controller that is separately installed and it could add new
+`apiGroup`s as they appear as `InfrastructurePublication`s in the remote
+cluster.
```

Comment:
Would this be something like https://github.com/crossplane/crossplane/issues/1637, but at the remote cluster?

Reply:
Not exactly, this separate controller would not only create the new roles but also bind them.

Reply:
Oh - I missed that this was granting access to the agent itself.